text/001-wasm-udfs.md

```diff
@@ -0,0 +1,348 @@
+- Feature Name: WASM UDFs
+- Start Date: 2022-10-31
+- RFC PR: [001](https://github.com/splitgraph/seafowl-rfcs/issues/001)
+- ClickUp task: [CU-3jv2jw7](https://app.clickup.com/t/3jv2jw7)
+- GH issue: [106](https://github.com/splitgraph/seafowl/issues/106)
+
+# Summary
+
+This RFC is the response to
+[#106](https://github.com/splitgraph/seafowl/issues/106). It describes a way to
+extend the existing support for WASM UDFs allowing for non-primitive data types
+such as strings in arguments and return values.
+
+# Motivation
+
+Seafowl already has the ability to introduce
+[custom UDFs](https://seafowl.io/docs/guides/custom-udf-wasm). These functions
+are capable of receiving a tuple of primitive values and returning a primitive.
+The currently available WASM primitives are integer and floating point numbers:
+`i32`, `i64`, `f32` and `f64`.
+
+While the opportunity to extend Seafowl is exciting, the current implementation
+falls short in several important ways:
+
+1. Seafowl's [data types](https://seafowl.io/docs/reference/types) are are much
+   richer than what can be trivially represented with scalar numeric primitives
+   (eg: `TEXT` and `DATE`). Most data types cannot currently be used as UDF
+   arguments or return values.
+1. Currently unsupported, future user-defined aggregate functions may receive an
+   array of tuples, which are also not easily represented with the numeric types
+   currently available (even if the tuples themselves contain only numbers).
+1. Aso currently unsupported user-defined table functions (the opposite of
+   aggregate functions) may return a list of tuples, which is also not possible
+   with the current limitations.
+1. Not a priority on the short term, but scalar functions with variable number
+   of arguments, eg: `CONCAT('prefix', table1.col1, 'suffix')` and those which
+   accept multiple types, eg: `MAX(0.5, 1)` need more flexibility than what can
+   be currently described in the `CREATE FUNCTION` expression.
+
+# Terminology
+
+- UDF - A _User Defined Function_ which can be used in `SELECT` statements just
+  like built-in functions.
+
+- WASI - WebAssembly System Interface - a standard set of functions which
+  essentially act as a `libc` for `wasm` programs.
+
+# Calling conventions
+
+A calling convention refers to the rules of invoking a function. In this case,
+it describes the following:
+
+1. _The semantics of function arguments and return values_. In order to receive
+   or return complex data types such as strings or arrays of tuples, the WASM
+   function must use its linear memory to communicate with the host process.
+   There are lots well-known calling conventions, for example, writing the
+   arguments to a segment of allocated WASM memory and passing the starting
+   pointer and length in bytes of the segment to the function as arguments.
+   Unfortunately, only a single value may be returned by functions, so this
+   method only works for passing input.
```

Comment:
The current implementation writes `stream length (i32) | messagepack data (char[])`.
This is more convenient than C structs because MessagePack deals with all the low-level serialization concerns.